### DISCUSSION TITLE
Tailwind CSS class sorting

### DISCUSSION THREAD

**Discussion by michaelhays** [biomejs/biome#164] [Q&A] [answered]
(Rome discussion: https://github.com/rome/tools/discussions/3672)

It would be great to have support for sorting Tailwind CSS classes, like https://github.com/tailwindlabs/prettier-plugin-tailwindcss.

For the time being, I'm using https://github.com/okonet/lint-staged, with the following config:

```js
const config = {
  '*.{js,json,jsx,ts,tsx}': [
    'prettier --write',
    'biome format --write',
    'biome lint --apply',
  ],
  '*.{css,html,md,mdx,yml,yaml}': ['prettier --write'],
}

export default config
```

This way, before committing, we first format committed files with Prettier (which sorts the Tailwind CSS classes with `prettier-plugin-tailwindcss`), and then format them again with Biome to make sure the formatting is consistent.

**Comment by ematipico** [comment]
There was some interest, but we would need to understand what's the logic of their formatting

**Reply by Gyurmatag** [reply]
When this will be implemented? This is the only one that blocks me to use this.

**Reply by ematipico** [reply]
@Gyurmatag when there's some interest, and someone wants to implement the logic. We're happy to help people with onboarding to the project.

**Comment by estubmo** [comment]
I'd be interested in working on this.

**Reply by Conaclos** [reply]
Thanks for your interest ❣️ We are still working on CSS support. We let you know once it is ready to bring this contribution. :)

**Reply by ematipico** [reply]
Maybe I misunderstood the plugin, I thought the plugin works on JSX files only?

**Reply by Conaclos** [reply]
Yes, you are right, Tailwind class sorting is not related to CSS support. @estubmo Thus you should be able to contribute :)

**Reply by estubmo** [reply]
Great. Would someone be able to onboard me?

**Reply by Conaclos** [reply]
I think we can add a lint rule for this feature.

First, I recommend reading our [contributing guide](https://github.com/biomejs/biome/blob/main/CONTRIBUTING.md) and [analyzer contributing guide](https://github.com/biomejs/biome/blob/main/crates/biome_analyze/CONTRIBUTING.md). You can use our [playground](https://biomejs.dev/playground/) (syntax window) to help to understand how the AST is shaped. Let's me know if you have other questions. You can also reach us in Discord.

I also recommend opening an issue (task) about this feature.

**Reply by ematipico** [reply]
I believe there's a misunderstanding. This should be a formatter feature, right?

**Reply by Conaclos** [reply]
@estubmo Which features of the plugin would you like to implement? Only sorting classes on the JSX `class` attribute? Because I am not a user of Tailwind, I really don't know if the provided options of the plugin are important to support...

Depending on what we want to implement we could determine if it is better to implement as a lint rule or as a formatter feature.

**Reply by estubmo** [reply]
My idea was just class sorting in jsx.

**Reply by Conaclos** [reply]
For now, I think we can stick with a lint rule. What do you think @ematipico?

**Reply by ematipico** [reply]
I agree 👍

**Reply by kharann** [reply]
Sorting inside of JSX is nice, but there's a lot of people using libraries as `tailwind-merge`, `classNames` and `class-variance-authority`. It's been quite nice to be able to sort in function calls and template literals. `prettier-plugin-tailwindcss` solved this by allowing, in the config, to specify what is a "tailwind function". Once the specified function wraps the node, it will be sorted.

In the original tailwind and plugin, everything that's prefixed by a specific keywords such as `class`, `className` will be sorted. Therefore I think a solution that looks for specific function names is more flexible than just looking inside JSX

**Reply by kharann** [reply]
Personally, I would love to be able to sort inside of my variables or use [cva](https://github.com/joe-bell/cva) to be able to create variants. Whaat do you think @estubmo @ematipico

**Reply by ematipico** [reply]
It seems we have a lot of information that we can use now, let's wait for the first implementation from @estubmo , that will be a good starting point

**Reply by DaniGuardiola** [reply]
@estubmo mind sharing the status of your work? i'd be happy to help, if that's useful to you. this is the only thing stopping me from switching to biome in all my projects.

**Comment by p7gg** [comment]
That would be really, really awesome to have.

**Comment by ImBIOS** [comment]
TailwindCSS class name sorting is one of the big reasons that keeps my company and myself in my project from adopting Biome and migrating from Prettier x ESLint.

**Reply by michaelhays** [reply]
FWIW, I'm still using the `lint-staged` approach described in my initial post, and it's been working well for me.

I know it may seem somewhat silly to use Biome at all for formatting if you still need Prettier to do the Tailwind sorting, but there's a meaningful difference in the "format on save" duration when switching from Prettier to Biome, which is a nice improvement for my workflow.

Obviously many of us would like to see this implemented directly in Biome, but I wouldn't let it stop you from trying it out.

**Reply by ImBIOS** [reply]
Woah, nice. If you say it's still worth trying, I planned to try it today with your biome + prettier (tailwind) configuration using lint-staged. Thanks for sharing the workaround and for the suggestion.

I normally use tailwind sorter using ESLint tho, so I might try your config using eslint tailwind sorter, in this way I can use className utilities.

**Comment by shrikantkalar023** [comment]
This would be pretty dope !!. Really like the prettier sorting of tailwind classes. Using Biome in all my backend projects. Once this is done I will also be able to shift fully to biome.

**Comment by kharann** [comment]
For those who havent noticed this might be in progres https://github.com/biomejs/biome/pull/1362

**Reply by DaniGuardiola** [reply]
it is

**Comment by ematipico** [comment, marked as answer]
Closing. Follow https://github.com/biomejs/biome/pull/1362 for updates

**Reply by DaniGuardiola** [reply]
i suggest "accepting as answer" for visibility, if that's an option

**Comment by MikeJakuszewski** [comment]
This would be great! Any idea when it would be implemented?

**Comment by Vladimir-Pirotskiy** [comment]
+++

**Comment by Isola92** [comment]
Currently moving a project to Tailwind. Would love to switch from Prettier but I cannot drop the class sorting.

**Reply by ematipico** [reply]
Have you read the answer to this discussion? The rule is there, even though it's a work in progress: https://github.com/biomejs/biome/pull/1362
